### ISSUE TITLE
How to use continiously?

### ISSUE THREAD

**Issue by olavt** [keras-rl/keras-rl#186]
I understand how to use the keras-rl framework in a limited train / test workflow as demonstrated in some of the samples.

But, how would one implement keras-rl in a scenario where one wants to deploy it into a real environment, where training starts from scratch and improves over time? Should I call "fit" and let in run for an episode and then persist the weights and then start over again? Or, how would the flow look like?

**Comment by pollini** [comment]
+1

**Comment by M00NSH0T** [comment]
In theory, yes, although I think you'd really want to pre-train this thing with historic data if you can. If your timesteps are say minutes, or even seconds, training live could take a pretty long time.  A lot of the benchmark programs that just do things like learn to play atari and whatnot can take hundreds of thousands or even millions of timesteps to reach their optimal policy.

Also, might want a different policy for running live vs training. Like you might want to just run a pure greedy policy live, but then have your training do something else, like epsilon-greedy or whatever so it doesn't get stuck learning a sub-optimal policy.

But yeah in general if you have a trained system, you'd just save the weights, then load them again when you're ready to use it live.

**Comment by olavt** [comment]
To give a little more context on the scenario. I want to use RL to control an air ventilation system with the goal of keeping air quality (as measured by CO2 level) at a reasonable level, while keeping energy cost down. Training data is not available, so it would need to train in production. The challenge as stated above is that it will take a long time to train and it will make lots of mistakes before it gets reasonably good at the job.

**Comment by M00NSH0T** [comment]
ok that might be doable, though I confess I haven't tried doing anything like this (i.e. training with purely live data) before, so definitely listen to others if they chime in here. I think you'd just need to use an off-policy algorithm, like DDPG or whatever as you may not be easily be able to break this into "episodes." The trick would be in getting your reward function defined correctly and maybe also set this up in a way that each snapshot can see a window of historic sequential data leading to the present since there may be a pretty big time delay between action and reward here.  The simple way is to put some time offset columns in your state (CO2 level now, one minute ago, two minutes ago, etc), or you can do something with recurrent neural nets / LSTM where you provide those past observations in a 3d array. I don't know the best way to set up your reward function off hand. You can go a couple different ways. Either everything's a penalty (i.e using electricity, exceeding CO2 threshold), or you can do something more complex. Either way, how you weight the cost of electricity vs CO2 levels will definitely play a role in the training speed, but the feedback loop with tweaking those might be too slow to be practical, so you'll probably just want to pick something reasonable and let it run.

I think one thing that could help with all this would be if you were able to pilot this idea in several buildings at once and do an asycn off policy method. Like A3C or D3PG/D4PG (which hopefully will be added to this repo soon). Then have each building be controlled by a different actor with some additional state variables involved about the overall system specs (system capacity, size of building, and maybe a one-hot vector to identify each building, etc.). That way you can have a critic that's learning a more general value function that informs the policy and hopefully it all could learn faster. I just don't know how ok it would be if this thing isn't working correctly in several buildings for a couple months while it learns.  You might need some backup system here, or have your environment force a fan on at some really poor CO2 threshold to ensure there's a backstop at least.

**Comment by olavt** [comment]
Thanks a lot for really good feedback! I have a simulated environment using keras-rl that I published here: https://github.com/olavt/gym_co2_ventilation. There I implemented a simple reward function for the scenario.

**Comment by guillem74** [comment]
@olavt @M00NSH0T So, is there any possibility to train during a lot of time and obtain some weights, and then update those weights with more training?

**Comment by olavt** [comment]
Yes, I have updated the environments here to include a more realistic training in production scenario. https://github.com/olavt/gym_co2_ventilation.

**Comment by guillem74** [comment]
@olavt Ok, so I think I do not understand the whole idea. When you fit() the second time after loading the weights, aren't you just overwriting the weights? Another thing is I don't understand the behavior of the memory in both the first and the second training.

**Comment by olavt** [comment]
@guillem74 No, the fit() method on the Agent does not overwrite the weights from a previous run. If you call fit(), save_weights() and then load_weights(), fit() it should continue the training from when you saved. I'm note sure how the memory of the agent should be preserved between runs. Maybe someone else could comment on that?

**Comment by pathway** [comment]
I was considering making a PR to build this into the keras-rl, its a small thing but makes training much easier.

**Comment by olavt** [comment]
@pathway Thanks, I have now added similar code to the test_keras_rl_continious.py example found here: https://github.com/olavt/gym_co2_ventilation

**Comment by pathway** [comment]
@olavt I updated my comment since it was first posted, it now reads the various buffers separately (see above).  As it was before, it was not loading the buffers.  I will make a PR when I have time.
